### ISSUE TITLE
Bug: Probably wrong fix of ESLint's prefer-template warning

### ISSUE THREAD

**Issue by pubmikeb** [eslint/eslint#15083]
### Environment

Node version: 16.9.1
npm version: 7.24.0
Local ESLint version: 8.0.0.beta.2
Global ESLint version: none
Operating System: Windows 10 21H1

### What parser are you using?

Other

### What did you do?

ESLint raises a warning `ESLint: Unexpected string concatenation.(prefer-template)`:

for the given code:

```js
res.setHeader("content-security-policy",
		"default-src 'self' https://*.google.com;"
		+ "script-src 'self' 'unsafe-eval' 'unsafe-inline' https://*.google.com;"
		+ "connect-src 'self' https://*.google.com;"
		+ "frame-ancestors 'none';"
		+ "report-to " + policyViolationReport
		+ ";");
```

### What did you expect to happen?

When I click on **ESLint: Fix: `prefere-template`** in IDE (or  `eslint --config eslintrc.json --fix app.js`), I expect to get:

```js
res.setHeader("content-security-policy",
		`"default-src 'self' https://*.google.com;"`
		+ `"script-src 'self' 'unsafe-eval' 'unsafe-inline' https://*.google.com;"`
		+ `"connect-src 'self' https://*.google.com;"`
		+ `"frame-ancestors 'none';"`
		+ `"report-to "${policyViolationReport};`);
```

### What actually happened?

The result in fact:

```js
res.setHeader("content-security-policy",
		`${"default-src 'self' https://*.google.com;"
		+ "script-src 'self' 'unsafe-eval' 'unsafe-inline' https://*.google.com;"
		+ "connect-src 'self' https://*.google.com;"
		+ "frame-ancestors 'none';"
		+ "report-to "}${  policyViolationReport
		 };`);
```

The possible issues are:
1. Unwanted space `${  policyViolationReport`
2. Unwanted line break for the closing clause ` };` of the last literal variable:

```js
+ "report-to "}${  policyViolationReport
 };`);
```

3. For some reason, `"default-src 'self' https://*.google.com;"` is represented as a placeholders (`${expression}`), while in fact it's just a text, not a variable in contrast to the `policyViolationReport`, which really should be represented as a literal placeholder — `${policyViolationReport}`.

4. BTW, the line break with literals can be implemented without concatenation symbols `+`, e.g.:

```js
`string text line 1
 string text line 2`
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

File for reproduction:
[app.js.txt](https://github.com/eslint/eslint/files/7192726/app.js.txt)

**Comment by nzakas** [comment]
Confirmed, this looks like a bug. I’m guessing the multi line string concatenation is messing with the rule logic.

**Comment by snitin315** [comment]
I will take it.

**Comment by mdjermanovic** [comment]
> 1. Unwanted space `${  policyViolationReport`

That's intentional because this rule doesn't know user's preferences about spacing in `${}`. This spacing can be automatically fixed in subsequent iterations by [template-curly-spacing](https://eslint.org/docs/rules/template-curly-spacing) and/or [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) rules.

> 2. Unwanted line break for the closing clause ` };` of the last literal variable:
> 
> ```js
> + "report-to "}${  policyViolationReport
>  };`);
> ```

That's intentional, it is assumed that the line breaks were there for readability.
 
> 3. For some reason, `"default-src 'self' https://*.google.com;"` is represented as a placeholders (`${expression}`), while in fact it's just a text, not a variable in contrast to the `policyViolationReport`, which really should be represented as a literal placeholder — `${policyViolationReport}`.

This looks like a bug.

> 4. BTW, the line break with literals can be implemented without concatenation symbols `+`, e.g.:
> 
> ```js
> `string text line 1
>  string text line 2`
> ```

It wouldn't be the same string value, because line breaks in template literals contribute to the value.

**Comment by mdjermanovic** [comment]
For this code:

> ```js
> res.setHeader("content-security-policy",
> 		"default-src 'self' https://*.google.com;"
> 		+ "script-src 'self' 'unsafe-eval' 'unsafe-inline' https://*.google.com;"
> 		+ "connect-src 'self' https://*.google.com;"
> 		+ "frame-ancestors 'none';"
> 		+ "report-to " + policyViolationReport
> 		+ ";");
> ```

I think the output should be:

> ```js
> res.setHeader("content-security-policy",
> 		"default-src 'self' https://*.google.com;"
> 		+ "script-src 'self' 'unsafe-eval' 'unsafe-inline' https://*.google.com;"
> 		+ "connect-src 'self' https://*.google.com;"
> 		+ "frame-ancestors 'none';"
> 		+ `report-to ${  policyViolationReport
> 		  };`);
> ```

**Comment by pubmikeb** [comment]
> This spacing can be automatically fixed in subsequent iterations by [template-curly-spacing](https://eslint.org/docs/rules/template-curly-spacing) and/or [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) rules.

In my case, both rules are enabled:

```json
"no-multi-spaces": "error",
"template-curly-spacing": "error"
```

But I still get the declared result. The `--fix` option is enabled in the IDEA settings.

**Comment by mdjermanovic** [comment]
> > This spacing can be automatically fixed in subsequent iterations by [template-curly-spacing](https://eslint.org/docs/rules/template-curly-spacing) and/or [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) rules.
> 
> In my case, both rules are enabled:
> 
> ```json
> "no-multi-spaces": "error",
> "template-curly-spacing": "error"
> ```
> 
> But I still get the declared result. The `--fix` option is enabled in the IDEA settings.

It works with ESLint's multipass autofix (`--fix`), please check the Fixed Code tab in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci10ZW1wbGF0ZTogZXJyb3IgKi9cbi8qIGVzbGludCBcIm5vLW11bHRpLXNwYWNlc1wiOiBcImVycm9yXCIgKi9cbi8qIGVzbGludCBcInRlbXBsYXRlLWN1cmx5LXNwYWNpbmdcIjogXCJlcnJvclwiICovXG5cbnJlcy5zZXRIZWFkZXIoXCJjb250ZW50LXNlY3VyaXR5LXBvbGljeVwiLFxuXHRcdFwiZGVmYXVsdC1zcmMgJ3NlbGYnIGh0dHBzOi8vKi5nb29nbGUuY29tO1wiXG5cdFx0KyBcInNjcmlwdC1zcmMgJ3NlbGYnICd1bnNhZmUtZXZhbCcgJ3Vuc2FmZS1pbmxpbmUnIGh0dHBzOi8vKi5nb29nbGUuY29tO1wiXG5cdFx0KyBcImNvbm5lY3Qtc3JjICdzZWxmJyBodHRwczovLyouZ29vZ2xlLmNvbTtcIlxuXHRcdCsgXCJmcmFtZS1hbmNlc3RvcnMgJ25vbmUnO1wiXG5cdFx0KyBcInJlcG9ydC10byBcIiArIHBvbGljeVZpb2xhdGlvblJlcG9ydFxuXHRcdCsgXCI7XCIpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=).

> When I click on **ESLint: Fix: `prefere-template`** in IDE

Maybe in this case your IDE doesn't run fixing through ESLint, but applies only that single fix. Each auto-fixable problem has an [EditInfo](https://eslint.org/docs/developer-guide/nodejs-api#-editinfo-type) property, which is the info on how to fix the code for that problem. Integrations can use this property to manually apply fixes if ESLint's built-in multipass autofix isn't what's needed for a particular use case.

**Comment by pubmikeb** [comment]
@mdjermanovic, should I rearrange the issue: extract all non-intentional behaviour (probably bugs) to the separate tickets and close this issue, which contains multiple issues?

P.S. I'll file an issue at IDEA bug tracker regarding the ESLint's built-in multipass autofix support.

**Comment by pubmikeb** [comment]
> It works with ESLint's multipass autofix (`--fix`), please check the Fixed Code tab in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci10ZW1wbGF0ZTogZXJyb3IgKi9cbi8qIGVzbGludCBcIm5vLW11bHRpLXNwYWNlc1wiOiBcImVycm9yXCIgKi9cbi8qIGVzbGludCBcInRlbXBsYXRlLWN1cmx5LXNwYWNpbmdcIjogXCJlcnJvclwiICovXG5cbnJlcy5zZXRIZWFkZXIoXCJjb250ZW50LXNlY3VyaXR5LXBvbGljeVwiLFxuXHRcdFwiZGVmYXVsdC1zcmMgJ3NlbGYnIGh0dHBzOi8vKi5nb29nbGUuY29tO1wiXG5cdFx0KyBcInNjcmlwdC1zcmMgJ3NlbGYnICd1bnNhZmUtZXZhbCcgJ3Vuc2FmZS1pbmxpbmUnIGh0dHBzOi8vKi5nb29nbGUuY29tO1wiXG5cdFx0KyBcImNvbm5lY3Qtc3JjICdzZWxmJyBodHRwczovLyouZ29vZ2xlLmNvbTtcIlxuXHRcdCsgXCJmcmFtZS1hbmNlc3RvcnMgJ25vbmUnO1wiXG5cdFx0KyBcInJlcG9ydC10byBcIiArIHBvbGljeVZpb2xhdGlvblJlcG9ydFxuXHRcdCsgXCI7XCIpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=).

@mdjermanovic, what does «multipass» mean in this context? Does it mean, that ESLint runs `--fix` recursively several times until no change is possible more?

[According](https://youtrack.jetbrains.com/issue/WEB-53345#focus=Comments-27-5364079.0-0) to the IDEA's team, `eslint --config eslintrc.json --fix reproduction.js` leads to the identical results as `eslint --fix` on Save in the IDE.

**Comment by pubmikeb** [comment]
I believe, I figured out the nature of the issue.

In case of fixing the issue via CLI, a proper result is received right after execution this *sole* command, while fixing an issue via IDEA UI requires more steps to be done manually: IDE returns an option with an extra space, which needed to be fixed separately.

Originally, the file was quite big and that's why there was a significant delay between fix and rescan. I assumed, that the result was a final-one, but in fact I should wait more time, until IDEA prepares the updated inspection outcomes.

> > 1. For some reason, `"default-src 'self' https://*.google.com;"` is represented as a placeholders (`${expression}`), while in fact it's just a text, not a variable in contrast to the `policyViolationReport`, which really should be represented as a literal placeholder — `${policyViolationReport}`.
> 
> This looks like a bug.

@mdjermanovic, since it's the only real issue in the case, should I extract this particular issue into a separate ticket and then this ticket can be closed?

**Comment by mdjermanovic** [comment]
> @mdjermanovic, since it's the only real issue in the case, should I extract this particular issue into a separate ticket and then #15083 can be closed?

It isn;t necessary, it's okay to keep this issue open for the problem we've all agreed on, and that is wrapping a starting sequence of string literals in `${}`. If you find some new problems related to this autofix, then it would be better to open a new issue.
